Emulated Wiimote Orientation Modifier

tenthmile opened the pull request:
A past patch from another user was not merged due to some flaws.

Therefore this is my attempt to implement this feature (https://bugs.dolphin-emu.org/issues/2948).

This patch works by adding a new attribute that is xor-ed with the existing sideway/upright bool.
Therefore no wxWidgets shenanigans occur.

I tried to keep it as flexible as possible so additional hotkeys could be used to modify other settings as well.

One thing to mention: Toggled buttons persist through save states (obviously as it's implemented in the ControllerEmu and has nothing to do with the game state). However due to this, it is sometimes confusing when it is active and when it's not.
Therefore I've added an Onscreen-Display-Message whenever toggled buttons are pressed (second commit of this pull-request).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dolphin-emu/dolphin/4145)
<!-- Reviewable:end -->

phire commented:
Yeah, this approach is much more sane.

lioncash commented:
Review status: 0 of 4 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[Source/Core/InputCommon/ControllerEmu.cpp, line 194 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrLb_WEJHj-jtijFXR:-KPrLb_XZOdwXuBu06Q3:blxu1i6) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/InputCommon/ControllerEmu.cpp#L194)):*
> ```cpp
> }
> 
> ControllerEmu::ModifySettingsButton::ModifySettingsButton(const std::string& _name) : Buttons(_name)
> ```

You can utilize the move constructor of std::string here like so:

```cpp
ControllerEmu::ModifySettingsButton::ModifySettingsButton(std::string name) : Buttons(std::move(name))
```

---

*[Source/Core/InputCommon/ControllerEmu.cpp, line 199 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrMgD_g7pkgbcth3hR:-KPrMgD_g7pkgbcth3hS:box1s37) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/InputCommon/ControllerEmu.cpp#L199)):*
> ```cpp
> }
> 
> void ControllerEmu::ModifySettingsButton::AddInput(const std::string& _name, bool toggle = false)
> ```

You should likely move the default parameter into the header, so it's a part of the interface. You can also use the move constructor here as well

```cpp
void ControllerEmu::ModifySettingsButton::AddInput(std::string name, bool toggle = false)
{
  controls.emplace_back(new ControlGroup::Input(std::move(name)));
...

```

---

*[Source/Core/InputCommon/ControllerEmu.h, line 225 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrNqntv4iSt_Lp0-37:-KPrNqnuSmead2VWkPpN:bisctfi) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/InputCommon/ControllerEmu.h#L225)):*
> ```objc
>   {
>   public:
>     ModifySettingsButton(const std::string& _name);
> ```

Try not to use prefixed underscores in variable names.

---

*[Source/Core/InputCommon/ControllerEmu.h, line 228 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrLyYrVoIY7c1sC8pT:-KPrLyYsoNZSg8DE7YzC:b-72cgou) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/InputCommon/ControllerEmu.h#L228)):*
> ```objc
>     void AddInput(const std::string& _name, bool toggle);
> 
>     void GetState()
> ```

You should likely move the implementation into the cpp file, so every file that includes ControllerEmu.h doesn't drag in the OnScreenDisplay stuff.

---

*[Source/Core/InputCommon/ControllerEmu.h, line 230 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrMEI9BB5rD_XgTHLF:-KPrMEI9BB5rD_XgTHLG:b-381bld) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/InputCommon/ControllerEmu.h#L230)):*
> ```objc
>     void GetState()
>     {
>       for (unsigned int i = 0; i < controls.size(); ++i)
> ```

`size_t`

---

*[Source/Core/InputCommon/ControllerEmu.h, line 235 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrMLTQPqN0eHILtcvI:-KPrMLTQPqN0eHILtcvJ:bkdhaly) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/InputCommon/ControllerEmu.h#L235)):*
> ```objc
> 
>         if (!associated_settings_toggle[i])  // not toggled
>           associated_settings[i] = state > numeric_settings[0]->GetValue();
> ```

This if statement should be braced.

---

*[Source/Core/InputCommon/ControllerEmu.h, line 246 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrMHqq6ILDKJQUIZqt:-KPrMHqrfI3UZaYKLS-h:b16ukyj) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/InputCommon/ControllerEmu.h#L246)):*
> ```objc
> 	      OSD::AddMessage(controls[i]->name + ": " + _trans("on"));
> 	    else
> 	      OSD::AddMessage(controls[i]->name + ": " + _trans("off"));
> ```

These lines are indented with tabs, but we use spaces.

---

*[Source/Core/InputCommon/ControllerEmu.h, line 255 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrNi-O1TgnNCjAMQjR:-KPrNi-O1TgnNCjAMQjS:btjcmg1) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/InputCommon/ControllerEmu.h#L255)):*
> ```objc
>     }
> 
>     std::vector<bool> associated_settings_toggle;  // is setting toggled or hold?
> ```

You may want to consider making these private and providing accessors for the data.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145)*
<!-- Sent from Reviewable.io -->

phire commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/HW/WiimoteEmu/WiimoteEmu.cpp, line 303 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrS6DQfNPP7TAT3snu:-KPrS6DRuSF3N1sHrGvx:b-on060y) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/Core/HW/WiimoteEmu/WiimoteEmu.cpp#L303)):*
> ```cpp
>   m_hotkeys->AddInput(_trans("Upright Toggle"), true);
>   // this setting modifier is not toggled
>   m_hotkeys->AddInput(_trans("Sideways Switch"), false);
> ```

It's not entire clear how "Switch" is different from "Toggle". They could both be interpreted as toggling actions, by a user.

Would naming it "Sideways Hold" work better, perhaps there is an even name I haven't thought of.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145)*
<!-- Sent from Reviewable.io -->

jloehr commented:
Wouldn't it be nicer if the hotkeys are associated with the respective BooleanSetting and are directly setting its value (It seems to expose a "SetValue"-Method). That way the Toggle and Switch/Hold will be reflected in the GUI and you don't need extra code in WiimoteEmu to get the actual value from all controls.

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-:-KPrqDKgzJmFPOGUd49E:b-y7g13l)*
<!-- Sent from Reviewable.io -->

phire commented:
@jloehr That's what the last PR did. 

It seems neat at the first glance, but it get really ugly real fast.

jloehr commented:
@phire Let me guess, changing the value by calling SetValue won't propagate to the GUI?

tenthmile commented:
@jloehr One reason why I don't like to mess with the BooleanSetting is that you interfere with the Config-File (e.g. when you keep the dialog open and later save it).

tenthmile commented:
I have a question about implementing the reviews (first time that I use github). Should I do a simple commit, or rebase for the pull-request?

jloehr commented:
@tenthmile Good point. IMO it's from a UX point of view not good, that a user can't check the current toggle state, without actually toggling it. That may confuse players, especially if the Controller Config Dialog Checkbox shows the wrong state.

phire commented:
Could we change the dialog text to say "Default Sideways" and "Default Upright" and move them out of Options? I think that would make things clearer. 

You probably need to add a small status thing to so users can see the current orientation of the wiimote and test out the new hotkeys.

As for github, we generally prefer fixups to be squashed into the relevant commits (aim for one commit per feature). But feel free to submit fixup commits for review and only squash them after everything is finalized.

tenthmile commented:
I've implemented the suggestions from the review.

@phire If I'm not mistaken, moving the settings out of the option group would break some config files?

phire commented:
That's probably true. I just think it would be nice to have all the orientation related options in a single section of the configuration window.

lioncash commented:
Review status: 1 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Core/InputCommon/ControllerEmu.cpp, line 215 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPsBYr3xoi39sINzgMH:-KPsBYr3xoi39sINzgMI:btxdxpa) ([raw file](https://github.com/dolphin-emu/dolphin/blob/0eef3b8b4c0e49fc8deb40876ea1901b3c19906a/Source/Core/InputCommon/ControllerEmu.cpp#L215)):*
> ```cpp
> 
>     if (!associated_settings_toggle[i])
>     {  // not toggled
> ```

Comments should be on their own line

---

*[Source/Core/InputCommon/ControllerEmu.cpp, line 219 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPsBde12-dHT3estFG5:-KPsBde12-dHT3estFG6:b-89a4l7) ([raw file](https://github.com/dolphin-emu/dolphin/blob/0eef3b8b4c0e49fc8deb40876ea1901b3c19906a/Source/Core/InputCommon/ControllerEmu.cpp#L219)):*
> ```cpp
>     }
>     else
>     {  // toggle (loading savestates does not en-/disable toggle)
> ```

Ditto

---

*[Source/Core/InputCommon/ControllerEmu.h, line 255 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrNi-O1TgnNCjAMQjR:-KPsBNaHwBonGIap5jyS:b8kdm80) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/InputCommon/ControllerEmu.h#L255)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> You may want to consider making these private and providing accessors for the data.
</details>
This still applies, considering they're only ever queried in external code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145)*
<!-- Sent from Reviewable.io -->

lioncash commented:
*[Source/Core/InputCommon/ControllerEmu.h, line 19 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPsCTHq9_U89EbKMumi:-KPsCTHrddPD0n4Rq0MD:b-45nssn) ([raw file](https://github.com/dolphin-emu/dolphin/blob/0eef3b8b4c0e49fc8deb40876ea1901b3c19906a/Source/Core/InputCommon/ControllerEmu.h#L19)):*
> ```objc
> #include "InputCommon/ControllerInterface/ControllerInterface.h"
> #include "InputCommon/GCPadStatus.h"
> #include "VideoCommon/OnScreenDisplay.h"
> ```

This include should go in the cpp file (Common.h likely doesn't need to be included).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145)*
<!-- Sent from Reviewable.io -->

leoetlino commented:
Review status: 1 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/Core/InputCommon/ControllerEmu.h, line 255 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrNi-O1TgnNCjAMQjR:-KPsCQZY_0qvGnVbZ7nv:birn5t8) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/InputCommon/ControllerEmu.h#L255)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This still applies, considering they're only ever queried in external code.
</details>
I think this should be `is setting toggled or held`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145)*
<!-- Sent from Reviewable.io -->

tenthmile commented:
@lioncash Sorry that I missed the accessor methods. Should be resolved now.

@phire  Regarding the Configuration Dialog and it's control positions. Should it be done in this pull-request or something for Dolphin 6.0?

tenthmile commented:
> Would naming it "Sideways Hold" work better, perhaps there is an even name I haven't thought of.

Done



> Review status: 1 of 4 files reviewed at latest revision, 4 unresolved discussions.
> 
> Source/Core/InputCommon/ControllerEmu.cpp, line 215 [r2] (raw file):
> 
>         if (!associated_settings_toggle[i])
>         {  // not toggled
> 
> Comments should be on their own line
> 
> Source/Core/InputCommon/ControllerEmu.cpp, line 219 [r2] (raw file):
> 
>         }
>         else
>         {  // toggle (loading savestates does not en-/disable toggle)
> 
> Ditto

Done and Done



> This include should go in the cpp file (Common.h likely doesn't need to be included).

Done (and you were right)



> You may want to consider making these private and providing accessors for the data.

Done (with both public attributes).
I've seen some cases in ControllerEmu.h where some accessors methods were implemented as one-liners. However the style guide states functions should be done otherwise. I sticked to the style guide, if that's wrong, I will change it.

tenthmile commented:
I pressed the "Update branch" button in the git-pull page. (Did not know what it does). But I reverted the change it did.
After that (@phire), I squashed all the commits for the feature.

tenthmile commented:
@lioncash @leoetlino @phire What do I need to do, to get this merged. I'm really confused about the silence... did I miss something github/reviewable.io related?

tenthmile commented:
Review status: 1 of 4 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/HW/WiimoteEmu/WiimoteEmu.cpp, line 303 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrS6DQfNPP7TAT3snu:-KQjP0g4_TtdNNGh-lNM:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/Core/HW/WiimoteEmu/WiimoteEmu.cpp#L303)):*
<details><summary><i>Previously, phire (Scott Mansell) wrote…</i></summary>
> It's not entire clear how "Switch" is different from "Toggle". They could both be interpreted as toggling actions, by a user.
> 
> Would naming it "Sideways Hold" work better, perhaps there is an even name I haven't thought of.
</details>
Done.

---

*[Source/Core/InputCommon/ControllerEmu.cpp, line 194 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrLb_WEJHj-jtijFXR:-KQjPZJQtubVSDW2eYwU:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/InputCommon/ControllerEmu.cpp#L194)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> You can utilize the move constructor of std::string here like so:
> 
> ```cpp
> ControllerEmu::ModifySettingsButton::ModifySettingsButton(std::string name) : Buttons(std::move(name))
> ```
</details>
Done.

---

*[Source/Core/InputCommon/ControllerEmu.cpp, line 199 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrMgD_g7pkgbcth3hR:-KQjPbDLl474niY08-kb:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/InputCommon/ControllerEmu.cpp#L199)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> You should likely move the default parameter into the header, so it's a part of the interface. You can also use the move constructor here as well
> 
> ```cpp
> void ControllerEmu::ModifySettingsButton::AddInput(std::string name, bool toggle = false)
> {
>   controls.emplace_back(new ControlGroup::Input(std::move(name)));
> ...
> 
> ```
</details>
Done.

---

*[Source/Core/InputCommon/ControllerEmu.cpp, line 215 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPsBYr3xoi39sINzgMH:-KQjP7RazeaOxfbL808a:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/0eef3b8b4c0e49fc8deb40876ea1901b3c19906a/Source/Core/InputCommon/ControllerEmu.cpp#L215)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Comments should be on their own line
</details>
Done.

---

*[Source/Core/InputCommon/ControllerEmu.cpp, line 219 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPsBde12-dHT3estFG5:-KQjP7rQbWxiVHGhh4r2:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/0eef3b8b4c0e49fc8deb40876ea1901b3c19906a/Source/Core/InputCommon/ControllerEmu.cpp#L219)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Ditto
</details>
Done.

---

*[Source/Core/InputCommon/ControllerEmu.h, line 225 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrNqntv4iSt_Lp0-37:-KQjPhkf1mrJCUtvhdL4:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/InputCommon/ControllerEmu.h#L225)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Try not to use prefixed underscores in variable names.
</details>
Done.

---

*[Source/Core/InputCommon/ControllerEmu.h, line 228 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrLyYrVoIY7c1sC8pT:-KQjPjCF4PNWtrORxd5i:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/InputCommon/ControllerEmu.h#L228)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> You should likely move the implementation into the cpp file, so every file that includes ControllerEmu.h doesn't drag in the OnScreenDisplay stuff.
</details>
Done.

---

*[Source/Core/InputCommon/ControllerEmu.h, line 230 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrMEI9BB5rD_XgTHLF:-KQjPjWCN2L9Z59XBCZs:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/InputCommon/ControllerEmu.h#L230)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> `size_t`
</details>
Done.

---

*[Source/Core/InputCommon/ControllerEmu.h, line 235 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrMLTQPqN0eHILtcvI:-KQjPmSw_YhBl2Lb7E60:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/InputCommon/ControllerEmu.h#L235)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This if statement should be braced.
</details>
Done.

---

*[Source/Core/InputCommon/ControllerEmu.h, line 246 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrMHqq6ILDKJQUIZqt:-KQjPk7seSQsym2F1vI5:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/InputCommon/ControllerEmu.h#L246)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> These lines are indented with tabs, but we use spaces.
</details>
Done.

---

*[Source/Core/InputCommon/ControllerEmu.h, line 255 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPrNi-O1TgnNCjAMQjR:-KQjPBOMvWGjEJB1WbCu:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/d4a63535406975eec1b2810068db6bda1dc971b9/Source/Core/InputCommon/ControllerEmu.h#L255)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary>
> I think this should be `is setting toggled or held`
</details>
Done.

---

*[Source/Core/InputCommon/ControllerEmu.h, line 19 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145#-KPsCTHq9_U89EbKMumi:-KQjP9LEBQN7Dwcjlsbe:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/0eef3b8b4c0e49fc8deb40876ea1901b3c19906a/Source/Core/InputCommon/ControllerEmu.h#L19)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This include should go in the cpp file (Common.h likely doesn't need to be included).
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145)*
<!-- Sent from Reviewable.io -->

tenthmile commented:
Reviewed 1 of 4 files at r1, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4145)*
<!-- Sent from Reviewable.io -->

phire commented:
What  this really needs now is somebody to test it.

JMC47 commented:
The settings themselves confuse me, but the toggles work as advertised in Super Paper Mario once I figured out how each option effected it.

Seems like the options themselves are a bit weird on whether the IR pointer shows up, but, that's not introduced by this build.

phire commented:
Yeah, pre-existing issues. Emulated Wiimotes could do with some TLC.
